wpengine/faustjs
# BH-731 WPMenu React component

- Bootstraps our `headless-components` package.
- Adds a `WPMenu` component to pull and display menu items from WP menu locations with GraphQL.
- Adds a reference implementation in the `example` folder to help us test components in use and create new ones.

<img width="585" alt="Screenshot 2020-11-30 at 16 19 31" src="https://user-images.githubusercontent.com/647669/100627898-e72a9200-3327-11eb-9fde-5fad24ac335e.png">

<img width="500" alt="Screenshot 2020-11-30 at 16 19 47" src="https://user-images.githubusercontent.com/647669/100627896-e5f96500-3327-11eb-900c-650e76219cd2.png">

## To test
1. Install and activate the [WPGraphQL plugin](https://wordpress.org/plugins/wp-graphql/) on your WordPress site.
2. Install and activate the wpe-headless plugin from this repo. (From your local WP root: `ln -s /Users/your-name/path/to/headless-framework/plugins/wpe-headless wpe-headless && wp activate wpe-headless`.)
3. In the WordPress admin area, add menu items to your Primary and Footer menu locations at Appearance → Menus.
4. Replace both instances of `http://dev.test` in `packages/headless-components/example/.env.local` with the path to your WordPress installation.
5. `cd packages/headless-components && npm install && npm run build`
6. `cd packages/headless-components/example && npm install && npm run dev`
7. Visit the Next.js site at http://localhost:3000.

You will see menu items from the 'primary' menu location in the header, and menu items from the 'footer' menu location in the footer.

## The WPMenu component
The menu component is defined at `packages/headless-components/src/components/wp-menu/index.js`.

Find the WPMenu component in use (twice) at `packages/headless-components/example/pages/index.js`. Menu data is loaded server-side in `getStaticProps()`:

```js
  await apolloClient.query({ query: WP_MENU_QUERY })
```

If you remove that, the menu will still load but data is fetched client-side instead of server-side.

By default, the menu loads items from the 'primary' WordPress menu location.

```js
<WPMenu />
```

Pass an alternative location with the `location` prop:

```js
<WPMenu location='footer' />
```

Default styling is imported in `packages/headless-components/example/pages/_app.js`.

## Future scope
- Pull out GraphQL requirement and use an adapter layer for data fetching once we better understand the scope of this.
- Improve default styling and accessibility. 
- Add hamburger/mobile-ready.button.
